src/datastore/connection.ts

```diff
@@ -1,8 +1,61 @@
 import { logError, parseArgBoolean, parsePort, stopwatch, timeout } from '../helpers';
 import * as postgres from 'postgres';
 import { isPgConnectionError } from './helpers';
+import { AsyncLocalStorage } from 'async_hooks';
 
-export type PgSqlClient = postgres.Sql<any>;
+export type PgSqlClient = postgres.Sql<any> | postgres.TransactionSql<any>;
+
+type UnwrapPromiseArray<T> = T extends any[]
+  ? {
+      [k in keyof T]: T[k] extends Promise<infer R> ? R : T[k];
+    }
+  : T;
+
+type SqlTransactionContext = {
+  usageName: string;
+};
+/**
+ * AsyncLocalStorage which determines if the current async context is running inside a SQL
+ * transaction.
+ */
+export const sqlTransactionContext = new AsyncLocalStorage<SqlTransactionContext>();
+
+/**
+ * Start a SQL transaction using a specific sql client. If this client was already scoped inside a
+ * `BEGIN` transaction, the callback will be called directly. This flexibility allows us to avoid
+ * repeating code while making sure we don't arrive at SQL errors such as
+ * `WARNING: there is already a transaction in progress` which may cause result inconsistencies.
+ * @param sql - SQL client
+ * @param callback - Callback with a scoped SQL client
+ * @param readOnly - If a `BEGIN` transaction should be marked as `READ ONLY`
+ * @returns Transaction results
+ */
+export async function sqlTransaction<T>(
+  sql: PgSqlClient,
+  callback: (sql: PgSqlClient) => T | Promise<T>,
+  readOnly = true
+): Promise<UnwrapPromiseArray<T>> {
+  if (sqlTransactionContext.getStore()) {
+    return callback(sql) as UnwrapPromiseArray<T>;
```

Comment:
I'm a bit uncertain about the typing here -- the return value of `callback` can be `T | Promise<T>`, but `sqlTransaction(...)` returns a Promise, so I think that makes certain usages of this function dependent on the js behavior where technically you can await something that isn't a promise, e.g.
```js
const test = 'abc';
console.log(await test);
// > 'abc'
```

That's probably why the typecast to `UnwrapPromiseArray` is required? Are there usages of the callback that are synchronous and actually need that behavior?